File: src/api/comparison.ts
import { invoke } from "@tauri-apps/api/core";
import type {
	ComparisonCursor,
	ComparisonHistoryDetail,
	ComparisonHistoryPage,
	SaveComparisonHistoryRequest,
} from "@/types/comparison";

/**
 * Persists one completed comparison and returns its history identifier.
 * @example await saveComparisonHistory({ query, results });
 */
const saveComparisonHistory = (request: SaveComparisonHistoryRequest) =>
	invoke<{ id: number }>("save_comparison_history", { request });

/**
 * Loads one bounded newest-first page without response bodies.
 * @example await listComparisonHistory(null, 30);
 */
const listComparisonHistory = (
	cursor: ComparisonCursor | null = null,
	limit = 30,
) =>
	invoke<ComparisonHistoryPage>("list_comparison_history", {
		request: { cursor, limit },
	});

/**
 * Loads one complete comparison for the history detail surface.
 * @example await getComparisonHistory(42);
 */
const getComparisonHistory = (id: number) =>
	invoke<ComparisonHistoryDetail>("get_comparison_history", {
		request: { id },
	});

export { getComparisonHistory, listComparisonHistory, saveComparisonHistory };

File: src-tauri/src/commands/comparison.rs
use crate::dto::comparison::{
    ComparisonDetailResponse, GetComparisonRequest, ListComparisonsRequest,
    ListComparisonsResponse, SaveComparisonRequest, SaveComparisonResponse,
};
use crate::error::IpcError;
use crate::services::comparison::ComparisonService;

/// Persists one completed comparison and returns its history identifier.
#[tauri::command]
pub(crate) async fn save_comparison_history(
    request: SaveComparisonRequest,
    service: tauri::State<'_, ComparisonService>,
) -> Result<SaveComparisonResponse, IpcError> {
    let id = service.save(request).await.map_err(IpcError::from)?;
    Ok(SaveComparisonResponse { id })
}

/// Returns one bounded newest-first page of comparison history.
#[tauri::command]
pub(crate) async fn list_comparison_history(
    request: ListComparisonsRequest,
    service: tauri::State<'_, ComparisonService>,
) -> Result<ListComparisonsResponse, IpcError> {
    service
        .list(request)
        .await
        .map(Into::into)
        .map_err(Into::into)
}

/// Returns one complete comparison for the history detail view.
#[tauri::command]
pub(crate) async fn get_comparison_history(
    request: GetComparisonRequest,
    service: tauri::State<'_, ComparisonService>,
) -> Result<ComparisonDetailResponse, IpcError> {
    service
        .find(request.id)
        .await
        .map(Into::into)
        .map_err(Into::into)
}
